File: src/steam/achievements.rs
use std::env;

use anyhow::{Context, Result};
use reqwest::Client;
use serde::{Deserialize, Serialize};

use super::{STEAM_ID, STEAM_TOKEN};

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SchemaForGame {
    game: Game,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct Game {
    #[serde(rename = "availableGameStats")]
    available_game_stats: GameStats,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GameStats {
    achievements: Vec<SchemaAchievement>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct SchemaAchievement {
    #[serde(rename = "displayName")]
    display_name: String,
    icon: String,
    description: String,
    name: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PlayerAchievements {
    #[serde(rename = "playerstats")]
    player_stats: PlayerStats,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PlayerStats {
    achievements: Option<Vec<PlayerAchievement>>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct PlayerAchievement {
    #[serde(rename = "apiname")]
    pub api_name: String,
    pub achieved: u32,
}

#[derive(Debug, PartialEq, Clone, Deserialize, Serialize)]
pub struct Achievement {
    pub api_name: String,
    pub achieved: bool,
    pub icon: String,
    pub display_name: String,
    pub description: String,
}

pub async fn fetch_game_achievements(
    app_id: u32,
    client: &Client,
) -> Result<Option<Vec<Achievement>>> {
    let steam_token = env::var(STEAM_TOKEN).context("fetching steam token env var failed")?;
    let steam_id = env::var(STEAM_ID).context("fetching steam id env var failed")?;

    let player_resp: reqwest::Response = client
        .get("https://api.steampowered.com/ISteamUserStats/GetPlayerAchievements/v0001/")
        .query(&[
            ("key", &steam_token),
            ("steamid", &steam_id),
            ("appid", &app_id.to_string()),
            ("format", &String::from("json")),
        ])
        .send()
        .await
        .context("getting response for player achievements failed")?;
    let player_resp_text = player_resp
        .text()
        .await
        .context("getting raw response text failed")?;
    if player_resp_text
        == r#"{"playerstats":{"error":"Requested app has no stats","success":false}}"#
    {
        return Ok(None);
    }
    let player_data: PlayerAchievements =
        serde_json::from_str(&player_resp_text).context(format!(
            "reading json failed from request to get achievements for {}: response: {}",
            app_id, player_resp_text
        ))?;
    let player_achievements = player_data.player_stats.achievements;
    if player_achievements.is_none() {
        return Ok(None);
    }

    let schema_resp: reqwest::Response = client
        .get("https://api.steampowered.com/ISteamUserStats/GetSchemaForGame/v2/")
        .query(&[
            ("key", &steam_token),
            ("appid", &app_id.to_string()),
            ("format", &String::from("json")),
        ])
        .send()
        .await
        .context("getting response for player achievements failed")?;
    let schema_resp_text = schema_resp
        .text()
        .await
        .context("getting raw response text failed")?;
    let schema_data: SchemaForGame = serde_json::from_str(&schema_resp_text).context(format!(
        "reading json failed from request to get achievements for {}: response: {}",
        app_id, schema_resp_text
    ))?;

    let mut achievements = vec![];
    for player_achievement in player_achievements.unwrap() {
        for schema_achievement in &schema_data.game.available_game_stats.achievements {
            if player_achievement.api_name == schema_achievement.name {
                achievements.push(Achievement {
                    api_name: player_achievement.api_name.to_owned(),
                    achieved: player_achievement.achieved == 1,
                    icon: schema_achievement.icon.to_owned(),
                    display_name: schema_achievement.display_name.to_owned(),
                    description: schema_achievement.description.to_owned(),
                })
            }
        }
    }

    Ok(Some(achievements))
}

File: internal/apis/steam/achievements.go
package steam

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"

	"go.mattglei.ch/lcp/internal/apis"
	"go.mattglei.ch/lcp/internal/secrets"
	"go.mattglei.ch/timber"
)

type playerAchievementsResponse struct {
	PlayerStats struct {
		Achievements *[]struct {
			ApiName    string `json:"apiname"`
			Achieved   int    `json:"achieved"`
			UnlockTime *int64 `json:"unlocktime"`
		}
	} `json:"playerStats"`
}

type schemaGameResponse struct {
	Game struct {
		GameStats struct {
			Achievements []struct {
				DisplayName string  `json:"displayName"`
				Icon        string  `json:"icon"`
				Description *string `json:"description"`
				Name        string  `json:"name"`
			} `json:"achievements"`
		} `json:"availableGameStats"`
	} `json:"game"`
}

type achievement struct {
	Achieved bool `json:"achieved"`
}

func fetchAchievementsPercentage(
	client *http.Client,
	appID int,
) (*float32, error) {
	params := url.Values{
		"key":     {secrets.ENV.SteamKey},
		"steamid": {secrets.ENV.SteamID},
		"appid":   {fmt.Sprint(appID)},
		"format":  {"json"},
	}
	resp, err := client.Get(
		"https://api.steampowered.com/ISteamUserStats/GetPlayerAchievements/v0001?" + params.Encode(),
	)
	if err != nil {
		return nil, fmt.Errorf(
			"%v sending request for player achievements from %d failed",
			err,
			appID,
		)
	}

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, fmt.Errorf(
			"%v reading response body for player achievements from %d failed",
			err,
			appID,
		)
	}
	if string(body) == `{"playerstats":{"error":"Requested app has no stats","success":false}}` {
		return nil, nil
	}
	if resp.StatusCode != http.StatusOK {
		timber.Warning(
			cacheInstance.LogPrefix(),
			"status code of",
			resp.StatusCode,
			"returned from API. Code of 200 expected from",
			resp.Request.URL.String(),
		)
		return nil, apis.ErrWarning
	}

	err = resp.Body.Close()
	if err != nil {
		return nil, fmt.Errorf("%w failed to close response body", err)
	}

	var playerAchievements playerAchievementsResponse
	err = json.Unmarshal(body, &playerAchievements)
	if err != nil {
		err = fmt.Errorf("%w failed to parse json for player achievements for %d", err, appID)
		timber.Debug("body:", string(body))
		return nil, err
	}

	if playerAchievements.PlayerStats.Achievements == nil {
		return nil, nil
	}

	params = url.Values{
		"key":    {secrets.ENV.SteamKey},
		"appid":  {fmt.Sprint(appID)},
		"format": {"json"},
	}
	req, err := http.NewRequest(
		http.MethodGet,
		"https://api.steampowered.com/ISteamUserStats/GetSchemaForGame/v2?"+params.Encode(),
		nil,
	)
	if err != nil {
		return nil, fmt.Errorf(
			"%v creating request for owned games failed for app id: %d",
			err,
			appID,
		)
	}
	gameSchema, err := apis.RequestJSON[schemaGameResponse](cacheInstance.LogPrefix(), client, req)
	if err != nil {
		return nil, fmt.Errorf("%w failed to get game schema for app id: %d", err, appID)
	}

	var achievements []achievement
	for _, playerAchievement := range *playerAchievements.PlayerStats.Achievements {
		for _, schemaAchievement := range gameSchema.Game.GameStats.Achievements {
			if playerAchievement.ApiName == schemaAchievement.Name {
				achievements = append(achievements, achievement{
					Achieved: playerAchievement.Achieved == 1,
				})
			}
		}
	}

	var totalAchieved int
	for _, achievement := range achievements {
		if achievement.Achieved {
			totalAchieved++
		}
	}
	achievementPercentage := (float32(totalAchieved) / float32(len(achievements))) * 100.0

	return &achievementPercentage, nil
}
